### PR TITLE
Use `cudf::thread_index_type` in `merge.cu`

### PR DESCRIPTION
**Pull request by divyegala**
## Description
This PR uses `cudf::thread_index_type` to avoid overflows.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — APPROVED** [review]

File: cpp/src/merge/merge.cu [hunk 1]
```diff
@@ -78,11 +78,13 @@ __global__ void materialize_merged_bitmask_kernel(
   size_type const num_destination_rows,
   index_type const* const __restrict__ merged_indices)
 {
-  size_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+  auto tid = cudf::thread_index_type{threadIdx.x} +
+             cudf::thread_index_type{blockIdx.x} * cudf::thread_index_type{blockDim.x};
```
Comment (vyasr) [inline cpp/src/merge/merge.cu:82]:
```suggestion
  auto tid = cudf::detail::grid_1d::global_thread_id();
```

File: cpp/src/merge/merge.cu [hunk 2]
```diff
@@ -99,8 +101,8 @@ __global__ void materialize_merged_bitmask_kernel(
     // Only one thread writes output
     if (0 == threadIdx.x % warpSize) { out_validity[word_index(destination_row)] = result_mask; }
 
-    destination_row += blockDim.x * gridDim.x;
-    active_threads = __ballot_sync(active_threads, destination_row < num_destination_rows);
+    tid += cudf::thread_index_type{blockDim.x} * cudf::thread_index_type{gridDim.x};
```
Comment (vyasr) [inline cpp/src/merge/merge.cu:104]: Maybe compute this once outside the loop and store as `stride`?

**Review by vyasr — CHANGES_REQUESTED** [review]
One requested change, one suggestion re: stride.

File: cpp/src/merge/merge.cu (hunk 2 above)
Comment (vyasr) [reply cpp/src/merge/merge.cu:104]: Also if you wait for #13996 to merge you can use that function instead here. At this point we're blocked on CI by #14018 anyway.

File: cpp/src/merge/merge.cu [hunk 3]
```diff
@@ -78,11 +78,13 @@ __global__ void materialize_merged_bitmask_kernel(
...
   index_type const* const __restrict__ merged_indices)
 {
-  size_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+  auto tid = detail::grid_1d::global_thread_id();
 
-  auto active_threads = __ballot_sync(0xffff'ffffu, destination_row < num_destination_rows);
+  auto active_threads = __ballot_sync(0xffff'ffffu, tid < num_destination_rows);
 
-  while (destination_row < num_destination_rows) {
+  auto stride = detail::grid_1d::grid_stride();
```
Comment (ttnghia) [inline cpp/src/merge/merge.cu:85]: `const`, and need to move it up to separate it from the other non-const vars.

**Review by vyasr — APPROVED** [review]

**Pull request merged**